### ISSUE TITLE
How to select specifial GPU for one screen?

### ISSUE THREAD

**Issue by gehaoyu-cn** [VirtualDrivers/Virtual-Display-Driver#110]
I have 2 GPUs and 2 vitrual screens, now 2 screens are working on single GPU, how to seperate 2 screens load into each GPU?

**Comment by itsmikethetech** [comment]
HDR release with GPU switching in testing now. https://github.com/itsmikethetech/Virtual-Display-Driver/releases/tag/24.7.18HDR

**Comment by matheusfaustino** [comment]
Can confirm the manual switching thing works. 

Basically, you need to find the name of your using the enum_dxgi (you can find in this thread https://github.com/nomi-san/parsec-vdd/issues/47) and then, download his release above, put the name of your GPU in adapter.txt file , put it inside `C:\IddSampleDriver` together with the options.txt and, install the driver (just delete the old driver, if you have it) .
It should select your GPU now for the screen virtual display. 

I think I will create a PR explaining this process in the README. The trick is that you need to do this before installing the driver, it is not hot-reload.

**Comment by zjoasan** [comment]
> it is not hot-reload.

Have you tried, just disable and re-enable the driver, instead of reinstalling it?

**Comment by matheusfaustino** [comment]
no, but from what I've seen in the code, I think it only gets the content from the adapter.txt when it is adding it (might be wrong here - didn't check the whole code)

https://github.com/itsmikethetech/Virtual-Display-Driver/blob/519936dbd9b677735fc7f64d76c9c1128ec00960/Virtual%20Display%20Driver%20(HDR)/IddSampleDriver/Driver.cpp#L139-L158

**Comment by bud3699** [comment]
> did not try it, but from what I've seen in the code, I think it only gets the content from the adapter.txt when it is adding it (might be wrong here - didn't check the whole code)
> 
> 
> 
> https://github.com/itsmikethetech/Virtual-Display-Driver/blob/519936dbd9b677735fc7f64d76c9c1128ec00960/Virtual%20Display%20Driver%20(HDR)/IddSampleDriver/Driver.cpp#L139-L158



You can just enable and re-enable the driver. It calls all the code again on driver initialise. Not install

**Comment by matheusfaustino** [comment]
Perfekt. Thanks for checking. Better than installing

**Comment by bud3699** [comment]
> Perfekt. Thanks for checking. Better than installing

No problem :) 

I already knew the answer so i would have been happy to share my knowledge. Better features to come soon enough!

**Comment by Nonary** [comment]
The latest changes in pre-release of this driver fixes the issue where it doesn't pick the proper GPU, but it will still be broken for Sunshine users due to how it is also mistakenly choosing the iGPU when attempting to capture the display.

Fortunately I fixed the issue on Sunshine, and I will be opening a PR later today.


Therefore, with the combination of these two changes, one from IDDSampleDriver side and one from sunshine, the issue where you can't stream in moonlight when using the virtual screen with more than one gpu will be resolved. No more black screens, also fixes the no video received from host error.

I'll comment here once I open that PR

**Comment by Nonary** [comment]
submitted https://github.com/LizardByte/Sunshine/pull/3002

**Comment by zjoasan** [comment]
@Nonary I glanced at your PR on Sunshines repo, although i don't know their repo. Is it likely that it's "our driver" that is "at fault" and perhaps we should some initial testing when grabbing GPU, and resasign if "blank". Or is it' Sunshine way of addressing our driver??

My thinking: How likely would it be for example OBS to experience the same? Or, I believe it was, Anydesk that had issues. Haven't tested any one of em. And I only have CPU based gpu (i7 gen4).

@bud3699 what do you think??

**Comment by Nonary** [comment]
> @Nonary I glanced at your PR on Sunshines repo, although i don't know their repo. Is it likely that it's "our driver" that is "at fault" and perhaps we should some initial testing when grabbing GPU, and resasign if "blank". Or is it' Sunshine way of addressing our driver??
> 
> My thinking: How likely would it be for example OBS to experience the same? Or, I believe it was, Anydesk that had issues. Haven't tested any one of em. And I only have CPU based gpu (i7 gen4).
> 
> @bud3699 what do you think??

The issues with the virtual display were mainly due to how Sunshine handled things, but the fix you implemented does help. One of the problems was that the driver might mistakenly attach to the iGPU, but this wasn’t as frequent as the Sunshine bug was.

We have a small utility program that tests DXGI against different adapters and monitors. If the test passes, the program assigns that GPU as the preferred one for capturing. The problem was that the utility only checked if we could duplicate the display in DXGI, not if it could actually capture frames. I’ve fixed that, so now Sunshine won’t encounter this issue anymore because IDDSampleDriver will fail if it can’t capture frames.

The reason we have this utility is due to a limitation with setting GPU preferences in DXGI—it can only be done once per process. So, we created a helper program to test DXGI and then report back to Sunshine. If the test passed, it would set the GPU preferences. The issue was that it was reporting success just because it could duplicate the display, without checking if it could capture frames. And since this can only be done once per process, Sunshine couldn’t recover or test the second GPU. By fixing the utility, the issue is now resolved.

**Comment by zjoasan** [comment]
@Nonary Thank you for the descriptive explanation, then we could add that bit later in the pipe perhaps. I'm a sucker for double checking and fallback solutions.

**Comment by fr1g** [comment]
How about in the newest beta? I've noticed that there's no longer any .txt file but a .xml file.

**Comment by itsmikethetech** [comment]
> How about in the newest beta? I've noticed that there's no longer any .txt file but a .xml file.

Yup! It can be edited in the XML or control app. :)
